### DISCUSSION TITLE
ArgoCD Parent Chart with separate dependencies installation.

### DISCUSSION THREAD

**Discussion by mjordan79** [argoproj/argo-cd#9421] [Q&A]
Hello, 
I have multiple microservices, each shipping with their own helm chart. I deploy these services separately in a continuous integration pipeline using a classic "helm upgrade --install" command.

I then have a parent chart, which include each of these microservices as a dependency (subcharts) which I want to install using ArgoCD. This is convenient because in this way I can have a single package in order to install all the microservices at once when I need it, out of the context of the CI pipeline.

Is this feasible / supported? I have noticed that the installations are somewhat colliding, but I'm not sure what is going wrong. Any advice? Should I play with "special" annotations / labels in order not to have collisions?

Sorry if this has been asked many times but I cannot find any specific guidelines about this.

**Comment by rishabh625** [comment]
since argocd does the helm install parent chart , subcharts gets installed , so it should be supported , didn't get u that installations are colliding?

**Reply by rishabh625** [reply]
update : argocd does  ``` helm dependency build ``` and ``` helm template ```

**Comment by nastacio** [comment]
It is unclear what kind of collision you are seeing, but I recently stumbled upon this blog entry covering Helm chart dependencies with Argo. Maybe the concrete examples can help you: https://cloud.redhat.com/blog/continuous-delivery-with-helm-and-argo-cd

**Comment by mjordan79** [comment]
@rishabh625 and @nastacio 
Thanks for your replies. Exactly as the article you've suggested, I'm installing my whole application using a parent chart and several subcharts as dependencies. On a first installation (clean environment) no problem. As the article says, doing a parent chart installation using Argo corresponds to the command:
helm template . <options> | kubectl apply -f - 
So the installation is done using kubectl and "helm ls" won't see any package installed.

But the microservices (and so the subcharts) evolve separately over time. So we have a CI process (in Jenkins) which package and push the subcharts along with the microservice, so they're installed independently from the other services, but using Helm this time. And here comes the problem (let's say "the collision": when I install one of these subcharts as an Helm installation, sometimes helm complaints saying the resource is already present but are not part of the installation.

I've read (correct me if I'm wrong) that I have to use the following annotations, in order to "import" existing resources into an helm installation (this is discussed here: https://github.com/helm/helm/pull/7649 )
meta.helm.sh/release-name and 
meta.helm.sh/release-namespace and
app.kubernetes.io/managed-by

I've put these annotations in my k8s artifacts, but when I install the whole application using the parent chart and ArgoCD, the values for these annotations are always overwritten with the application name (in all artifacts and for all subcharts, regardless of the service being installed, preventing me from doing a correct management). The managed-by annotation is populated with the string "Helm" in both cases, so no issue here. Hence the conflict I was mentioning in this thread when I switch back to helm while installing the subcharts separately:

```
rendered manifests contain a resource that already exists. Unable to
continue with install: Service "abc" in namespace "xyz" exists and
cannot be imported into the current release: invalid ownership
metadata; 

<snip>
```

In practice, I want to install all my subcharts at once using Argo and a parent chart, but I also want to overwrite correctly these subcharts using Helm and the individual subchart for the installation.

I mean, subcharts are still independent charts for which I would like to continue to have the freedom to install them separately using Helm. 
I would like to know if there is some good practice to solve this issue. Is there some options for Argo that allow to control how these annotations are handled / populated?
